### DISCUSSION TITLE
Packaging Quarto for Nix

### DISCUSSION THREAD

**Discussion by winniehell** [quarto-dev/quarto-cli#11118] [Q&A]
### Description

I'm using the [NixOS package](https://github.com/NixOS/nixpkgs/blob/66ec997b38ce5bcbbc8769ff5247144a1b400ccc/pkgs/development/libraries/quarto/default.nix) to install Quarto. As of writing this, it uses the system installed versions of Quarto's dependencies instead of downloading new copies matching the versions in [configuration](https://github.com/quarto-dev/quarto-cli/blob/71945532e1fc1a5cf113117f6d5ff5bee3991797/configuration). As far as I understand [@mcanouil's comment](https://github.com/quarto-dev/quarto-cli/issues/11117#issuecomment-2423963125), that approach is not supported. My understanding of the configuration file is that only a single version (rather than a version range) is supported by Quarto. Does that mean, providing all dependencies specifically for Quarto (and therefore potentially causing duplicates on the system) is the only option?

**Comment by mcanouil** [comment]
Quarto does embed the dependencies.
The person doing the NixOS package needs to do the same or something similar to ensure the `configuration` file is respected.

**Reply by mcanouil** [reply]
Note that this is not specific to NixOS. FreeBSD, conda, etc. need to do the same.

Currently, Quarto natively build what you can find in the assets of each releases: https://github.com/quarto-dev/quarto-cli/releases

- https://github.com/quarto-dev/quarto-cli/discussions/4480

**Comment by cscheid** [comment]
> My understanding of the configuration file is that only a single version (rather than a version range) is supported by Quarto

That's correct. I don't know how (whether?) NixOS supports configuration pinning, but Quarto depends quite narrowly on the versions of the OS: Deno, for its variations in the Deno standard object and its typing, and Pandoc, for its variations in output generation. If there's a mismatch, it's almost certain that Quarto won't function properly.

To be clear, we do not maintain the NixOS Quarto package ourselves. Unfortunately, if that package (or the environment) is not using the correct versions, then we cannot offer troubleshooting support.

**Comment by winniehell** [comment]
>  I don't know how (whether?) NixOS supports configuration pinning

Me neither because I have only started using NixOS and my own understanding is still very limited. For the time being, I may just stick with the development version (which was easy to get started with! :heart:) for local use and find a way to build a Docker image for CI/CD.

Thank you for the fast and helpful input here! :+1:

**Comment by winniehell** [comment]
more discussion is happening here: https://github.com/NixOS/nixpkgs/issues/349444

**Comment by mcanouil** [comment]
> I don't know how (whether?) NixOS supports configuration pinning

For reference, Nixpkgs does not.

**Reply by winniehell** [reply]
from my limited knowledge, the best option would be to have a [flake](https://nixos.wiki/wiki/flakes) that reads version numbers from the `configuration` file. if Quarto maintainers don't object, the `flake.nix` could live in this repository which would make easier to find.

**Reply by mcanouil** [reply]
The issue is not the location, it's the maintenance burden.
NixOS is far from being a major distribution.
Adding a Flake or whatever means someone knows how to works and can fix issues in it.
Quarto has a limited team thus a limited time to spend on this maintenance.

Carlos (the lead developer/maintainer) explained that in various issues/discussions including the one about upgrading to Deno 2.0 related to the issue from NixOS package manager repository.

If there is support for NixOS is likely needs to come from the NixOS community and be maintained by it, but it needs to follow Quarto dependencies configuration.

**Reply by winniehell** [reply]
My question was rather: would it be possible to put the nix / flake configuration in the Quarto repository and add a note that it is not officially supported by the Quarto team? I think that would make maintenance easier.

I would be more than happy to contribute the necessary files but I still need to do research on how to do it, so I can't give an estimate on when that will happen.

**Reply by cscheid** [reply]
> My question was rather: would it be possible to put the nix / flake configuration in the Quarto repository and add a note that it is not officially supported by the Quarto team?

It's technically possible but we won't do it until we are ready to support it. We take the official repository seriously and we are already spread out too thin in the promises we make here. A file in the official repository is eventually a promise that we will support it, because it's in the official repository.

**Reply by winniehell** [reply]
yes, I understand :+1:

let me also be clear that I see no responsibility for you to contribute in any way to the Nix packaging. you provide multiple ways to get Quarto running locally and sufficient instructions. your fast responses here and in the other nix-related issues are already above my expectations, thank you!
